packages/lodestar/src/api/impl/validator/index.ts

```diff
@@ -277,16 +296,43 @@ export function getValidatorApi({chain, config, logger, metrics, network, sync}:
       const startSlot = computeStartSlotAtEpoch(epoch);
       await waitForSlot(startSlot); // Must never request for a future slot > currentSlot
 
+      const nextEpoch = chain.clock.currentEpoch + 1;
       const state = await chain.getHeadStateAtCurrentEpoch();
-
       const duties: routes.validator.ProposerDuty[] = [];
       const indexes: ValidatorIndex[] = [];
 
-      // Gather indexes to get pubkeys in batch (performance optimization)
-      for (let i = 0; i < SLOTS_PER_EPOCH; i++) {
-        // getBeaconProposer ensures the requested epoch is correct
-        const validatorIndex = state.getBeaconProposer(startSlot + i);
-        indexes.push(validatorIndex);
+      // We allow requesting proposal duties only one epoch in the future
+      // Note: There is a small probability that returned validators differs
+      // than what is returned when the epoch is reached.
+      if (epoch === nextEpoch) {
+        const cachedDutiesForEpoch = nextEpochProposerDutyCache.get(nextEpoch);
+
+        if (cachedDutiesForEpoch) {
+          indexes.push(...cachedDutiesForEpoch);
+        } else {
+          const futureProposers = computeProposersForEpoch(state, nextEpoch);
+
+          indexes.push(...futureProposers);
+
+          // Do not keep previous future proposal duties, so clear cache
+          // before setting a new value.
+          nextEpochProposerDutyCache.clear();
+          nextEpochProposerDutyCache.set(nextEpoch, futureProposers);
+        }
+      } else {
+        // Gather indexes to get pubkeys in batch (performance optimization)
+        for (let i = 0; i < SLOTS_PER_EPOCH; i++) {
+          // getBeaconProposer ensures the requested epoch is correct
+          const validatorIndex = state.getBeaconProposer(startSlot + i);
+          indexes.push(validatorIndex);
+        }
+
+        for (const cachedEpoch of nextEpochProposerDutyCache.keys()) {
+          // Do not keep past cached future proposal duties.
+          if (cachedEpoch < epoch) {
```

Comment:
in this case, epoch is current epoch, maybe we even don't need the cache if `cachedEpoch === currentEpoch`, so `<=` instead?

Reply:
I think using `<=` would mean that if next epoch is computed, and then if the current epoch is requested and after that, the next epoch is requested again, it would have to be computed again (because requesting the current epoch would have deleted it)

Reply:
>  if next epoch is computed, and then if the current epoch is requested and after that, the next epoch is requested again, it would have to be computed again (because requesting the current epoch would have deleted it)

I think it's not deleted because `cachedEpoch` is the next epoch so `cachedEpoch <= epoch` is `false` (because `epoch` is `current epoch`)

Reply:
oh I mean in the case of a malicious actor, who sees that during the current epoch N, the first call to epoch N+1 is expensive before caching, and hence during the same epoch N, continues to make request for epoch N+1, and then epoch N (to delete the previous cache) and then epoch N + 1 again, in other to overwhelm the system

Reply:
You are right. It should be fine to delete the proposers in the cache at the current epoch (the cache value that was previous computed as at a future epoch) when the current epoch is finally reached. I'll update the code now to reflect this.